src/main/java/it/smartcommunitylab/aac/webauthn/controller/WebAuthnAuthenticationController.java

```diff
@@ -0,0 +1,117 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.server.ResponseStatusException;
+
+import io.swagger.v3.oas.annotations.Hidden;
+import it.smartcommunitylab.aac.webauthn.model.WebAuthnAssertionResponse;
+import it.smartcommunitylab.aac.webauthn.model.WebAuthnAuthenticationStartRequest;
+import it.smartcommunitylab.aac.webauthn.model.WebAuthnLoginResponse;
+import it.smartcommunitylab.aac.webauthn.provider.WebAuthnIdentityService;
+import it.smartcommunitylab.aac.webauthn.service.WebAuthnRpService;
+import it.smartcommunitylab.aac.webauthn.service.WebAuthnRpServiceRegistrationRepository;
+
+/**
+ * Manages the endpoints connected to the authentication ceremony of WebAuthn.
+ * 
+ * The authentication ceremony is used to authenticate the client by verifying
+ * that it owns a credential that was previously registered on this server
+ * (thanks to the {@link WebAuthnRegistrationController}).
+ */
+@Controller
+@RequestMapping
+public class WebAuthnAuthenticationController {
+
+    @Autowired
+    private WebAuthnRpServiceRegistrationRepository webAuthnRpServiceReigistrationRepository;
+
+    /**
+     * Serves the page to start a new WebAuthn authentication ceremony.
+     * Ensure the request mapping value matches the one returned from
+     * {@link WebAuthnIdentityService#getLoginUrl}
+     */
+    @Hidden
+    @RequestMapping(value = "/auth/webauthn/login/{providerId}", method = RequestMethod.GET)
+    public String authenticatePage() {
+        return "webauthn/authenticate";
+    }
+
+    /**
+     * Starts a new WebAuthn authentication ceremony by generating a new Credential
+     * Request Options object and returning it to the user.
+     * 
+     * The challenge and the information about the ceremony are temporarily stored
+     * in the session.
+     */
+    @Hidden
+    @PostMapping(value = "/auth/webauthn/assertionOptions/{providerId}", consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
+    @ResponseBody
+    public WebAuthnLoginResponse generateAssertionOptions(@RequestBody @Valid WebAuthnAuthenticationStartRequest body,
+            @PathVariable("providerId") String providerId) {
+        try {
+            WebAuthnRpService rps = webAuthnRpServiceReigistrationRepository.get(providerId);
+
+            String username = body.getUsername();
+            if (!isValidUsername(username)) {
+                throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Invalid username");
```

Comment:
Got it. In this case, however, throwing an IllegalArgumentException would cause the server to return a response with status code '500: Internal server error'. Since the username is a value the user inputs, semantically we should return a '400: bad request'.